### PR TITLE
Incentives component

### PR DESCRIPTION
**Pull request by zwkohn**
## Overview
This adds an Incentives component to the library
Screenshot 2023-08-07 at 4 12 23 PM

## Security

REMINDER: All file contents are public.

- [ ] I have ensured no secure credentials or sensitive information remain in code, metadata, comments, etc.
  - [ ] Please verify that you double checked that .storybook/preview.js does not contain your participant access key details. 
  - [ ] There are no temporary testing changes committed such as API base URLs, access tokens, print/log statements, etc.
- [ ] These changes do not introduce any security risks, or any such risks have been properly mitigated.

Describe briefly what security risks you considered, why they don't apply, or how they've been mitigated.

### Testing
- [ ] MyDataHelps iOS app
- [ ] MyDataHelps Android app
- [ ] [MyDataHelps website](mydatahelps.org)

### Documentation
- [ ] I have added relevant Storybook updates to this PR as well.
- [ ] If this feature requires a developer doc update, tag @CareEvolution/api-docs.

Consider "Squash and merge" as needed to keep the commit history reasonable on `main`.

## Reviewers

Assign to the appropriate reviewer(s). Minimally, a second set of eyes is needed ensure no non-public information is published. Consider also including:
- Subject-matter experts
- Style/editing reviewers
- Others requested by the content owner

### CODE REVIEW

**Comment by chrisnowak** [comment]
Few thoughts:
- I think we want to name this a lot more narrowly; e.g. "IncentivesProgressBar"; or honestly just "ProgressBar" since it may not always be in the context of incentives but could work for goal setting etc.
- We should do a bit of a review of other incentives progress bars we have done for other studies...e.g. PROGRESS...may suggest different options or look/feel.  PROGRESS may be the gold standard here in some ways
image - I would see if we can track a bit more to that look / feel
- Would suggest we do not have an "increment" or anything...but rather we should leave this as a presentational component and have:
  - background (string that does into CSS background prop; could be color or gradient)
  - fill (string that does into CSS background prop of the fill layer; could be color or gradient)
  - fillPercent (how wide the fill bar is): 0 to 100
  - steps[] e.g. the placement of the various icons...each of these would have
    - stepPercent: number 0 to 100 for where the step should be placed on the progress bar
    - a react element for the thing that should be displayed on the step...could be an icon or something...
image

Then I think maybe another component for "IncentiveProgressBarStepIcon" that has some options like a $$ amount that you can use in the "step" array...

**Comment by ajmenca** [comment]
A couple of reference points from recent work on AoU Rewards.  They use two variations of this kind of idea; one shows progress "in" level progression, the other shows progress across the entire reward program.

image
image

image
image

- I think that for both the AoU and Progress designs, having the right side of the fill bar be curved looks really nice, I think we should integrate that.
- I think both of these designs also animate the bar when the percentage changes; that's definitely a feature we should have. There's also some subtle animations that the AoU Bar has; if you hit 100%, the bar animates to the top, and *then* the star icon changes from gray to orange.  This is also a really nice effect that looks polished.
- In your screenshots, it looks like some things might need to get adjusted left/right by a pixel or so.  There's some aliasing between the circle of the "100" and the blue border, and also in the second screenshot there's some white pixels leaking through on the left side of the completion bar.

**Comment by bxh** [comment]
AOU clearly has more than one kind of progress bar in mind.

Screenshot 2023-08-15 160615

**Comment by Suvarna-Thejas-CE** [comment]
I like the idea of making this just a pure progress bar. I created an issue #162 that could be a wrapper around this component that has labels/text that make it useful for a specific use case. 

This one I think would be good with the parameters that Chris suggested above.

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 1]
```diff
@@ -0,0 +1,50 @@
+.progress-bar {
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:1]: Should prefix all CSS classes with .mdhui-, so this should be .mdhui-progress-bar

I would also nest all classes in this file under that - so for the fill it shuold be ```.mdhui-progress-bar .mdhui-progress-bar-fill```

Comment (chrisnowak) [reply src/components/presentational/ProgressBar/ProgressBar.css:1]: (alternative to consider at some point is using CSS modules; but so far we haven't taken that leap and CSS is still being overridden in places)

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	display:inline-block;
+	margin: 10px 0px;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:3]: Would remove the margin on this

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	width: 100%;
+	max-width: 500px;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:5]: Why a max-width?  Will look awkward on desktop so I would remove

File: src/components/presentational/ProgressBar/ProgressBar.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import "./ProgressBar.css";
+import React, { ReactElement, useEffect } from "react";
+
+export interface ProgressBarProps {
+	background?: string;
+	fill?: string;
+	fillPercent: number;
+	steps?: [
+		{
+			stepPercent: number;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.tsx:10]: "step" prefix here seems unnecessary since this will be in a "steps" array; so would just go with percent/icon

File: src/components/presentational/ProgressBar/ProgressBar.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import "./ProgressBar.css";
+import React, { ReactElement, useEffect } from "react";
+
+export interface ProgressBarProps {
+	background?: string;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.tsx:5]: We use color suffix for color properties, so would use backgroundColor and fillColor.

Also would make both background/fill a ColorDefinition and not a string - that was probably introduced after this PR but see the button component for an example (ColorDefinition allows specifying 2 colors, one for light mode and one for dark mode)

File: src/components/presentational/ProgressBar/ProgressBar.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	fillPercent: number;
+	steps?: [
+		{
+			stepPercent: number;
+			stepIcon: ReactElement;
+		}
+	]
+}
+
+ProgressBar.defaultProps = {
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.tsx:16]: What the heck how have I never heard about "defaultProps" before now

File: src/components/presentational/ProgressBar/ProgressBar.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	background: "#d3d3d3",
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.tsx:17]: Would have this be the closest CSS backround color var from the globalCss file

File: src/components/presentational/ProgressBar/ProgressBar.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import "./ProgressBar.css";
+import React, { ReactElement, useEffect } from "react";
+
+export interface ProgressBarProps {
+	background?: string;
+	fill?: string;
+	fillPercent: number;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.tsx:7]: I think we can add this component to the view builder just by itself, but that requires a couple additional props.

We need to add the "innerRef" prop so this can be dragged/dropped, and I also would support a "defaultMargin" prop that we would set to true for the view builder which would add the default 16px margin to this (see the link for example)

File: src/components/presentational/ProgressBar/ProgressBar.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+
+export default {
+	title: "Presentational/ProgressBar",
+	component: ProgressBar,
+	parameters: {
+		layout: 'auto',
+	},
+	argTypes: {
+		background: { control: 'color', description: "The background color of the progress bar" },
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.stories.tsx:16]: Documentation!  Awesome.  Need to start doing that...

File: src/components/presentational/ProgressBar/ProgressBar.tsx (hunk 5 above)
Comment (chrisnowak) [reply src/components/presentational/ProgressBar/ProgressBar.tsx:5]: Oh hmm, i'm realizing that fill might actually be a gradient...yet let me think about that for a sec

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	position: absolute;
+	z-index: 2;
+	display:inline-block;
+}
+
+.progress-bar .step-icon {
+	position: absolute;
+	transform: translate(-100%, 0)
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:43]: This does make things a bit awkward if you have a stop near the beginning of the progress bar - it will get clipped out...there's no way to have a stop at 0 reliably I think (since steps are based on percent).

That being said, this seems like the best and most reasonable solution to this and, as demonstrated by the stories, works pretty well.  I don't think having a stop right at the beginning makes a ton of sense

Comment (chrisnowak) [reply src/components/presentational/ProgressBar/ProgressBar.css:43]: and you could always fix that by having the element at the beginning have a left margin or something

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+.progress-bar .step-icon span {
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:46]: I wouldn't assume a span will be used as the child element...

So there is actually some cool stuff you did in the story here for the various little star icons and how they are colored etc.

I would not just waste that in a story but would actually invest in a separate component called "ProgressBarStep"...it could be exported under the progress bar...such that we have some out of box progress bar steps that can be placed on this (but someone could also come up with their own to supply to this).

and the props could be like:
- borderColor?: ColorDefinition
- backgroundColor: ColorDefinition
- children  (this is where you'd put in the font awesome icon for the star, or the text 75 or whatnot)
- height (based on this it would auto-calculate the top margin to center itself vertically, e.g. for the large star)


e.g. for the big star at the end your step would be like:

```
steps: [{
		stepPercent: 100,
		stepIcon: <ProgressBarStep borderColor="yellow" backgroundColor="orange" height="24px">
<FontAwesomeIcon icon={faStar} style={{ color: "yellow" }} />
</ProgressBarStep>
	}],
```

Basically to have to limit the amount of work someone has to do to get it to look like the storybook, while still allowing some flexibility

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	--icon_border_width: 2px;
+}
+
+.mdhui-progress-bar .mdhui-progress-bar-default-margin {
+	margin: 16px;
+}
+
+.mdhui-progress-bar .mdhui-progress-bar-background {
+	border-radius: 500px;
+	border-width: var(--icon_border_width) ;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:15]: Huh...did you just add a border to this?  I didn't think there was a border before

File: src/components/presentational/ProgressBar/ProgressBar.css [hunk 13]
```diff
@@ -0,0 +1,51 @@
+.mdhui-progress-bar {
+	display: block;
+	width: 100%;
+	position: relative;
+	--icon_size: 24px;
+	--icon_border_width: 2px;
+}
+
+.mdhui-progress-bar .mdhui-progress-bar-default-margin {
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.css:9]: Looks like this isn't working because you have a space between these and this class is applied to the progress bar itself.

Also, you will need to adjust the width like we do for buttons so it's ```width:calc(100% - 32px)```, otherwise it overflows

File: src/components/presentational/ProgressBar/ProgressBar.css (hunk 12 above)
Comment (chrisnowak) [reply src/components/presentational/ProgressBar/ProgressBar.css:15]: i would probably drop the border because it makes it look a bit awkward when stuff overlaps it:
Screenshot 2023-11-30 at 2 04 36 PM

File: src/components/presentational/ProgressBar/ProgressBar.stories.css [hunk 14]
```diff
@@ -0,0 +1,3 @@
+.sb-story {
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.stories.css:1]: What does this CSS do?

File: src/components/presentational/ProgressBar/ProgressBar.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+	},
+	argTypes: {
+		backgroundColor: { control: 'object', description: "The background color of the progress bar" },
+		fillColor: { control: 'object', description: "The fill color of the progress bar" },
+		fillPercent: { control: 'number', description: "The percent of the progress bar that is filled" },
+		steps: { control: 'object', description: "An array of steps to display on the progress bar. A step specifies the position on the progress bar as a percent and a React element to display at that position." },
+	}
+} as Meta<typeof ProgressBar>;
+
+const Template: StoryFn<typeof ProgressBar> = (args: ProgressBarProps) => <Layout colorScheme="auto"><ProgressBar {...args} /></Layout>;
```
Comment (chrisnowak) [inline src/components/presentational/ProgressBar/ProgressBar.stories.tsx:25]: Minor, maybe put "defaultMargin" on this so it's spaced out from the edges

**Review by chrisnowak — APPROVED** [review]
Looks good!

File: src/components/presentational/ProgressBar/ProgressBar.stories.css (hunk 14 above)
Comment (zwkohn) [reply src/components/presentational/ProgressBar/ProgressBar.stories.css]: Trying to get some padding in the storybook component. I replaced this with an extra div in the Template
```
const Template: StoryFn<typeof ProgressBar> = (args: ProgressBarProps) => <Layout colorScheme="auto"><div style={{padding: "35px 0"}}><ProgressBar {...args} /></div></Layout>;
```
